### ISSUE TITLE
How to Precompute Mean

### ISSUE THREAD

**Issue by tonyma163** [narger-ef/FHE-BERT-Tiny#2]
Hello sir,

May I know how did you extract the precomputed mean? Since I am trying to replicate your idea to generate the mean files using sst2 training datasets, but my results do not match your mean files.

**Comment by lorenzorovida** [comment]
Hello there,

the precomputed mean was obtained by using the SST-2 training dataset... we run the circuit for each training sentenc and observed the mean and variance values. Then, they were "averaged"

**Comment by tonyma163** [comment]
Since I noticed the mean files contain 55 non zero values, so I guess it would be calculating the mean for each token position. But how to compute the mean for each token position over their hidden features (128)?

Do you have a notebook or program extracting these means, so I could have a better understanding.

**Comment by lorenzorovida** [comment]
I found an old code (which is definitely not a nice code ^^) that may help you

```
sentences = [
    "The sun sets behind the mountains, casting a golden glow.",
    "A cat lazily stretches on the windowsill, basking in the sunlight.",
    ...]

mean_distribution = []
var_distribution = []

for sentence in sentences:
    text = "[CLS] " + sentence + " [SEP]"

    tokenized = tokenizer(text)
    tokenized_text = tokenizer.tokenize(text)
    segments_ids = [1] * len(tokenized_text)
    indexed_tokens = tokenizer.convert_tokens_to_ids(tokenized_text)
    tokens_tensor = torch.tensor([indexed_tokens])
    segments_tensors = torch.tensor([segments_ids])
    
    xx = model.embeddings(tokens_tensor, segments_tensors)
    xx = model.encoder.layer[0].attention.self(xx)[0].double()
    
    w_output_dense = model.encoder.layer[0].attention.output.dense.weight.clone().detach().double().transpose(0, 1)
    b_output_dense = model.encoder.layer[0].attention.output.dense.bias.clone().detach().double()

    xx = torch.matmul(xx, w_output_dense) + b_output_dense
    xx = xx + model.embeddings(tokens_tensor, segments_tensors)

    means = []
    variances = []

    for i in range(12):
        xi = xx.squeeze()[i]
        means.append(torch.mean(xi.squeeze()).item())
        variances.append(1 / math.sqrt(torch.var(xi.squeeze()).item()))
    
    mean_distribution.append(np.array(means))
    var_distribution.append(np.array(variances))
```

**Comment by tonyma163** [comment]
I get it, thank you!!

**Comment by tonyma163** [comment]
I have tried my best to replicate your method, and I have compared my precomputed variables with yours and found out they are not the same except for the first mean, var, and vy variables which are very close to your files.

```
Precision of mean_0_0: 0.9999998748343766
Precision of mean_0_1: 0.9963048022094373 <- NOT accurate 
Precision of mean_1_0: 0.9939214921803704 <- NOT accurate 
Precision of mean_1_1: 0.9976584139486315 <- NOT accurate 
Precision of var_0_0: 0.9999999932322889
Precision of var_0_1: 0.9965047249215336 <- NOT accurate 
Precision of var_1_0: 0.9960562611711485 <- NOT accurate 
Precision of var_1_1: 0.9966372052437913 <- NOT accurate 
Precision of vy_0_0: 0.9999999932322889
Precision of vy_0_1: 0.9965047249215335 <- NOT accurate 
Precision of vy_1_0: 0.9960562611711484 <- NOT accurate 
Precision of vy_1_1: 0.9966372048907628 <- NOT accurate 
Precision of normbias_0_0: 1.0
Precision of normbias_0_1: 1.0
Precision of normbias_1_0: 1.0
Precision of normbias_1_1: 1.0
```

I thought the 0.996 accuracy may enough for the experiment so I also tried using those variables to test the main.cpp and it show "The decryption failed because the approximation error is too high."

So I think my code exists some bugs there. Could you help me to have a look? Orz

```
# Encoder 1
mean_distribution_0_0 = []
var_distribution_0_0 = []
vy_distribution_0_0 = [[] for _ in range(128)]
normbias_distribution_0_0 = []

mean_distribution_0_1 = []
var_distribution_0_1 = []
vy_distribution_0_1 = [[] for _ in range(128)]
normbias_distribution_0_1 = []

for sentence in tqdm(valid_dataset['sentence']):
    text = "[CLS] " + sentence + " [SEP]"

    # Tokenization
    [ ... ]
    # Embeddings
    x = model.bert.embeddings(tokens_tensor, segments_tensors)
    original_input_tensor = x.double()
    input_tensor = x.double()

    # Self-Attention
    fin = model.bert.encoder.layer[0].attention.self(x)[0].double()

    w_output_dense = model.bert.encoder.layer[0].attention.output.dense.weight.clone().detach().double().transpose(0, 1)
    b_output_dense = model.bert.encoder.layer[0].attention.output.dense.bias.clone().detach().double()

    fin2 = torch.matmul(fin, w_output_dense) + b_output_dense
    fin2_backup = fin2.clone()
    fin2_backup = fin2_backup + original_input_tensor

    mean_0_0 = []
    var_0_0 = []

    fin3_whole = []
    for i in range(len(original_input_tensor.squeeze())):
        fin2 = fin2_backup.squeeze()[i]

        current_mean = torch.mean(fin2.squeeze()).item()
        epsilon = model.bert.encoder.layer[0].attention.output.LayerNorm.eps
        current_var = 1 / math.sqrt(torch.var(fin2.squeeze()).item() + epsilon)

        # save mean and variance
        mean_0_0.append(np.array(current_mean))
        var_0_0.append(np.array(current_var))

        fin3_corr = (fin2.squeeze() - current_mean) * current_var

        w_output_layernorm = model.bert.encoder.layer[0].attention.output.LayerNorm.weight.clone().detach().double().unsqueeze(0)
        b_output_layernorm = model.bert.encoder.layer[0].attention.output.LayerNorm.bias.clone().detach().double()

        # calculate vy
        vy = w_output_layernorm.squeeze(0) * current_var
        
        # save vy & normbias
        normbias = b_output_layernorm
        vy_distribution_0_0[i].append(np.array(vy))
        normbias_distribution_0_0.append(np.array(normbias))

        fin3_corr = fin3_corr * w_output_layernorm + b_output_layernorm
        fin3_whole.append(fin3_corr)

    mean_distribution_0_0.append(np.array(mean_0_0))
    var_distribution_0_0.append(np.array(var_0_0))

    fin3_whole = torch.cat(tuple(fin3_whole), 0).unsqueeze(0)
    input_tensor = fin3_whole
    fin_4 = torch.matmul(fin3_whole, model.bert.encoder.layer[0].intermediate.dense.weight.transpose(0, 1).double()) + model.bert.encoder.layer[0].intermediate.dense.bias

    fin_5 = torch.nn.functional.gelu(fin_4)
    fin_6 = torch.matmul(fin_5, model.bert.encoder.layer[0].output.dense.weight.transpose(0, 1).double()) + model.bert.encoder.layer[0].output.dense.bias
    fin_6 = fin_6 + fin3_whole

    mean_0_1 = []
    var_0_1 = []

    fin7_whole = []
    for i in range(len(input_tensor.squeeze())):
        fin7 = fin_6.squeeze()[i]

        current_mean = torch.mean(fin7.squeeze()).item()
        epsilon = model.bert.encoder.layer[0].output.LayerNorm.eps
        current_var = 1 / math.sqrt(torch.var(fin7.squeeze()).item() + epsilon)

        # save mean and variance
        mean_0_1.append(np.array(current_mean))
        var_0_1.append(np.array(current_var))

        fin7_corr = (fin7.squeeze() - current_mean) * current_var

        w_output_layernorm = model.bert.encoder.layer[0].output.LayerNorm.weight.clone().detach().double().unsqueeze(0)
        b_output_layernorm = model.bert.encoder.layer[0].output.LayerNorm.bias.clone().detach().double()

        # calculate vy
        vy = w_output_layernorm.squeeze(0) * current_var

        # save vy & normbias
        normbias = b_output_layernorm
        vy_distribution_0_1[i].append(np.array(vy))
        normbias_distribution_0_1.append(np.array(normbias))

        fin7_corr = fin7_corr * w_output_layernorm + b_output_layernorm
        fin7_whole.append(fin7_corr)
        
    mean_distribution_0_1.append(np.array(mean_0_1))
    var_distribution_0_1.append(np.array(var_0_1))

    fin7_whole = torch.cat(tuple(fin7_whole), 0).unsqueeze(0)
    original_input_tensor = fin7_whole
    input_tensor = fin7_whole
```

**Comment by tonyma163** [comment]
I don't know why it suddenly works :D
